### PR TITLE
Improve the logic of handling old protocol versions

### PR DESCRIPTION
**Pull request by mzhangmzz**
This PR introduces two changes:
1. Complete the logic that make nodes panic if their protocol version is old. Previously, the check was only triggered at epoch boundaries. If validators restart their nodes, their nodes will keep running. 
2. Introduce a new protocol feature that forbids blocks produced with old protocol versions. This can help kick out validators that are not upgraded.

### CODE REVIEW

File: chain/chain/src/chain.rs [hunk 1]
```diff
@@ -1134,6 +1135,26 @@ impl Chain {
...
+            self.runtime_adapter.get_epoch_protocol_version(header.epoch_id())
+        {
+            if checked_feature!(
+                "protocol_feature_reject_blocks_with_outdated_protocol_version",
+                RejectBlocksWithOutdatedProtocolVersions,
+                epoch_protocol_version
+            ) {
+                if header.latest_protocol_version() < epoch_protocol_version {
+                    error!(
+                        "header protocol version {} epoch protocol version {}",
```
Comment (mm-near) [inline chain/chain/src/chain.rs:1149]: you're missing "is smaller than"

File: chain/chain/src/chain.rs [hunk 2]
```diff
@@ -2136,6 +2157,12 @@ impl Chain {
             return Err(Error::Orphan);
         }
 
+        let epoch_protocol_version =
+            self.runtime_adapter.get_epoch_protocol_version(block.header().epoch_id())?;
+        if epoch_protocol_version > PROTOCOL_VERSION {
+            panic!("The client protocol version is older than the protocol version of the network. Please update nearcore");
```
Comment (mm-near) [inline chain/chain/src/chain.rs:2163]: I'd also print the epoch_protocol_version and PROTOCOL_VERSION in this 'panic'

File: chain/client/src/client.rs [hunk 3]
```diff
@@ -465,6 +465,13 @@ impl Client {
             .runtime_adapter
             .get_epoch_id_from_prev_block(&head.last_block_hash)
             .expect("Epoch hash should exist at this point");
+        let protocol_version = self
+            .runtime_adapter
+            .get_epoch_protocol_version(&epoch_id)
+            .expect("Epoch info should be ready at this point");
+        if protocol_version > PROTOCOL_VERSION {
+            panic!("The client protocol version is older than the protocol version of the network. Please update nearcore");
```
Comment (mm-near) [inline chain/client/src/client.rs:473]: as above -- consider printing the protocol_version and PROTOCOL_VERSION in the message.

File: integration-tests/src/tests/client/process_blocks.rs [hunk 4]
```diff
@@ -103,14 +103,16 @@ pub fn create_nightshade_runtimes(genesis: &Genesis, n: usize) -> Vec<Arc<dyn Ru
...
     env: &mut TestEnv,
     blocks_number: u64,
     height: BlockHeight,
+    protocol_version: ProtocolVersion,
 ) -> BlockHeight {
     let next_height = height + blocks_number;
     for i in height..next_height {
-        let block = env.clients[0].produce_block(i).unwrap().unwrap();
+        let mut block = env.clients[0].produce_block(i).unwrap().unwrap();
+        block.mut_header().set_lastest_protocol_version(protocol_version);
```
Comment (mm-near) [inline integration-tests/src/tests/client/process_blocks.rs:115]: typo in lastest ?

File: chain/chain-primitives/src/error.rs [hunk 5]
```diff
@@ -146,6 +146,9 @@ pub enum Error {
     /// `next_bps_hash` doens't correspond to the actual next block producers set
     #[error("Invalid Next BP Hash")]
     InvalidNextBPHash,
+    /// The block has a protocol version that's outdated
```
Comment (mm-near) [inline chain/chain-primitives/src/error.rs:149]: let's also update CHANGELOG.md

File: chain/chain-primitives/src/error.rs [hunk 6]
```diff
@@ -146,6 +146,9 @@ pub enum Error {
...
+    #[error("Invalid protocol version")]
+    InvalidProtocolVersion,
```
Comment (mm-near) [inline chain/chain-primitives/src/error.rs:151]: nit: everywhere else we call it 'outdated' not 'invalid'

**Review by mm-near — APPROVED** [review]

File: chain/chain-primitives/src/error.rs (hunk 5 above)
Comment (mzhangmzz) [reply chain/chain-primitives/src/error.rs:149]: Do I add the feature in CHANGELOG.md when we stabilize it?

File: chain/chain-primitives/src/error.rs (hunk 6 above)
Comment (mzhangmzz) [reply chain/chain-primitives/src/error.rs:151]: Yeah but all the other errors starts with InvalidXXX... I think I'll stick with Invalid

**Pull request merged**